File: app/messages.py
from dataclasses import dataclass
from enum import Enum
from typing import Dict, List, Self, Tuple, Type, TypeAlias, cast

from app.message_elements import (
    ChainHashElement,
    EncodedShortChannelIdsElement,
    GlobalFeaturesElement,
    LocalFeaturesElement,
    MessageTypeElement,
    PointElement,
    RemainderElement,
    SerializedElement,
    ShortChannelIDElement,
    SignatureElement,
    SingleByteElement,
    U16Element,
    U16VarBytesElement,
    U32Element,
    U64Element,
)


class MessageProperty(Enum):
    """
    An ElementKey represents the unique names for datatypes which are encoded in documentation, for example see [channel_announcement](https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-channel_announcement-message) in Bolt 7.
    """

    TYPE = "type"
    GLOBAL_FEATURES = "global_features"
    LOCAL_FEATURES = "local_features"
    NUM_PONG_BYTES = "num_pong_bytes"
    PING_OR_PONG_BYTES = "ping_or_pong_bytes"
    SIGNATURE = "signature"
    NODE_SIGNATURE_1 = "node_signature_1"
    NODE_SIGNATURE_2 = "node_signature_2"
    BITCOIN_SIGNATURE_1 = "bitcoin_signature_1"
    BITCOIN_SIGNATURE_2 = "bitcoin_signature_2"
    CHANNEL_FEATURES = "channel_features"
    CHAIN_HASH = "chain_hash"
    SHORT_CHANNEL_ID = "short_channel_id"
    NODE_ID_1 = "node_id_1"
    NODE_ID_2 = "node_id_2"
    BITCOIN_KEY_1 = "bitcoin_key_1"
    BITCOIN_KEY_2 = "bitcoin_key_2"
    FIRST_TIMESTAMP = "first_timestamp"
    TIMESTAMP_RANGE = "timestamp_range"
    ENCODED_SHORT_CHANNEL_IDS = "encoded_short_ids"
    FULL_INFORMATION = "full_information"
    TIMESTAMP = "timestamp"
    MESSAGE_FLAGS = "message_flags"
    CHANNEL_FLAGS = "channel_flags"
    CLTV_EXPIRY_DELTA = "cltv_expiry_delta"
    HTLC_MINIMUM_MSAT = "htlc_minimum_msat"
    FEE_BASE_MSAT = "fee_base_msat"
    FEE_PROPORTIONAL_MILLIONTHS = "fee_proportional_millionths"
    HTLC_MAXIMUM_MSAT = "htlc_maximum_msat"
    REMAINDER = "remainder"
    FIRST_BLOCK_NUM = "first_block_num"
    NUMBER_OF_BLOCKS = "number_of_blocks"
    SYNC_COMPLETE = "sync_complete"


KeyedElement: TypeAlias = Tuple[MessageProperty, Type[SerializedElement]]
MessagePropertiesDict: TypeAlias = Dict[MessageProperty, SerializedElement]


@dataclass
class Message:
    id: int
    name: str
    properties: MessagePropertiesDict

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return [
            (MessageProperty.TYPE, MessageTypeElement),
        ]

    @classmethod
    def from_bytes(cls, data: bytes):
        properties = {}
        chunked = bytes(data)
        for key, feature in cls.features():
            el, chunked = feature.from_bytes(bytes(chunked))
            properties[key] = el
            if feature is MessageTypeElement:
                el = cast(MessageTypeElement, el)
                id: int = el.id
                name: str = el.name
        if len(chunked) > 0:
            properties[MessageProperty.REMAINDER], chunked = RemainderElement.from_bytes(chunked)
        assert len(chunked) == 0, f"Unexpected data left: {chunked}"
        return cls(id, name, properties)  # pyright: ignore

    def to_bytes(self) -> bytes:
        data = b""
        for key, feature in self.features():
            data += self.properties[key].to_bytes()
        if MessageProperty.REMAINDER in self.properties:
            data += self.properties[MessageProperty.REMAINDER].to_bytes()
        return data

    def __str__(self):
        out = []
        for key, property in self.properties.items():
            out.append(f"{key.value}: {property}")
        return f"{self.__class__.__name__}({', '.join(out)})"

    def __getattr__(self, name: str):
        """Fallback for dynamically created properties."""
        try:
            return self.properties[MessageProperty[name.upper()]]
        except KeyError:
            raise AttributeError(f"{name} not found")

    @property
    def type_id(self):
        return cast(MessageTypeElement, self.type).id

    @property
    def type_element(self):
        return cast(MessageTypeElement, self.type).name

    @property
    def type_name(self):
        return self.name

    @property
    def length(self):
        return len(self.to_bytes())


class InitMessage(Message):
    id = 16

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.GLOBAL_FEATURES, U16VarBytesElement),
            (MessageProperty.LOCAL_FEATURES, U16VarBytesElement),
        ]

    @property
    def global_features(self):
        return cast(GlobalFeaturesElement, self.properties[MessageProperty.GLOBAL_FEATURES])

    @property
    def local_features(self):
        return cast(LocalFeaturesElement, self.properties[MessageProperty.LOCAL_FEATURES])


class PingMessage(Message):
    id: int = 18
    name: str = "ping"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.NUM_PONG_BYTES, U16Element),
            (MessageProperty.PING_OR_PONG_BYTES, U16VarBytesElement),
        ]

    @property
    def num_pong_bytes(self):
        return cast(U16Element, self.properties[MessageProperty.NUM_PONG_BYTES])

    @classmethod
    def create(cls, num_pong_bytes: int, message: bytes):
        return PingMessage(
            id=18,
            name="ping",
            properties={
                MessageProperty.TYPE: MessageTypeElement(id=18, name="ping"),
                MessageProperty.NUM_PONG_BYTES: U16Element(num_bytes=num_pong_bytes),
                MessageProperty.PING_OR_PONG_BYTES: U16VarBytesElement(len(message), data=message),
            },
        )


class PongMessage(Message):
    id = 19
    name = "pong"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.PING_OR_PONG_BYTES, U16VarBytesElement),
        ]

    @property
    def ping_or_pong_bytes(self):
        return cast(U16VarBytesElement, self.properties[MessageProperty.PING_OR_PONG_BYTES])

    @property
    def num_bytes(self):
        return self.ping_or_pong_bytes.num_bytes

    @classmethod
    def create_from_ping(cls, msg: PingMessage):
        return PongMessage(
            id=19,
            name="pong",
            properties={
                MessageProperty.TYPE: MessageTypeElement(id=19, name="pong"),
                MessageProperty.PING_OR_PONG_BYTES: U16VarBytesElement(
                    msg.num_pong_bytes.num_bytes,
                    data=b"0" * msg.num_pong_bytes.num_bytes,
                ),
            },
        )


class ChannelAnnouncementMessage(Message):
    id = 256
    name = "channel_announcement"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.NODE_SIGNATURE_1, SignatureElement),
            (MessageProperty.NODE_SIGNATURE_2, SignatureElement),
            (MessageProperty.BITCOIN_SIGNATURE_1, SignatureElement),
            (MessageProperty.BITCOIN_SIGNATURE_2, SignatureElement),
            (MessageProperty.CHANNEL_FEATURES, U16VarBytesElement),
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.SHORT_CHANNEL_ID, ShortChannelIDElement),
            (MessageProperty.NODE_ID_1, PointElement),
            (MessageProperty.NODE_ID_2, PointElement),
            (MessageProperty.BITCOIN_KEY_1, PointElement),
            (MessageProperty.BITCOIN_KEY_2, PointElement),
        ]

    @property
    def node_signature_1(self):
        return cast(SignatureElement, self.properties[MessageProperty.NODE_SIGNATURE_1])

    @property
    def node_signature_2(self):
        return cast(SignatureElement, self.properties[MessageProperty.NODE_SIGNATURE_2])

    @property
    def bitcoin_signature_1(self):
        return cast(SignatureElement, self.properties[MessageProperty.BITCOIN_SIGNATURE_1])

    @property
    def bitcoin_signature_2(self):
        return cast(SignatureElement, self.properties[MessageProperty.BITCOIN_SIGNATURE_2])

    @property
    def channel_features(self):
        return cast(U16VarBytesElement, self.properties[MessageProperty.CHANNEL_FEATURES])

    @property
    def chain_hash(self):
        return cast(ChainHashElement, self.properties[MessageProperty.CHAIN_HASH])

    @property
    def short_channel_id(self):
        return cast(ShortChannelIDElement, self.properties[MessageProperty.SHORT_CHANNEL_ID])

    @property
    def node_id_1(self):
        return cast(PointElement, self.properties[MessageProperty.NODE_ID_1])

    @property
    def node_id_2(self):
        return cast(PointElement, self.properties[MessageProperty.NODE_ID_2])

    @property
    def bitcoin_key_1(self):
        return cast(PointElement, self.properties[MessageProperty.BITCOIN_KEY_1])

    @property
    def bitcoin_key_2(self):
        return cast(PointElement, self.properties[MessageProperty.BITCOIN_KEY_2])


class ChannelUpdateMessage(Message):
    id = 258
    name = "channel_update"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        # https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-channel_update-message
        return super().features() + [
            (MessageProperty.SIGNATURE, SignatureElement),
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.SHORT_CHANNEL_ID, ShortChannelIDElement),
            (MessageProperty.TIMESTAMP, U32Element),
            (MessageProperty.MESSAGE_FLAGS, SingleByteElement),
            (MessageProperty.CHANNEL_FLAGS, SingleByteElement),
            (MessageProperty.CLTV_EXPIRY_DELTA, U16Element),
            (MessageProperty.HTLC_MINIMUM_MSAT, U64Element),
            (MessageProperty.FEE_BASE_MSAT, U32Element),
            (MessageProperty.FEE_PROPORTIONAL_MILLIONTHS, U32Element),
            (MessageProperty.HTLC_MAXIMUM_MSAT, U64Element),
        ]


class GossipTimestampFilterMessage(Message):
    id = 265
    name = "gossip_timestamp_filter"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.FIRST_TIMESTAMP, U32Element),
            (MessageProperty.TIMESTAMP_RANGE, U32Element),
        ]

    @property
    def chain_hash(self):
        return cast(ChainHashElement, self.properties[MessageProperty.CHAIN_HASH])

    @property
    def first_timestamp(self):
        return cast(U32Element, self.properties[MessageProperty.FIRST_TIMESTAMP])

    @property
    def timestamp_range(self):
        return cast(U32Element, self.properties[MessageProperty.TIMESTAMP_RANGE])


class GossipChannelIDQueryFlags(Enum):
    CHANNEL_ANNOUNCEMENTS = 0
    CHANNEL_UPDATES_NODE_1 = 1
    CHANNEL_UPDATES_NODE_2 = 2
    NODE_ANNOUNCEMENTS_NODE_1 = 3
    NODE_ANNOUNCEMENTS_NODE_2 = 4


class QueryShortChannelIDsMessage(Message):
    id = 261
    name = "query_short_channel_ids"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.ENCODED_SHORT_CHANNEL_IDS, EncodedShortChannelIdsElement),
        ]

    @property
    def chain_hash(self):
        return cast(ChainHashElement, self.properties[MessageProperty.CHAIN_HASH])

    @property
    def encoded_short_channel_ids(self):
        return cast(
            EncodedShortChannelIdsElement,
            self.properties[MessageProperty.ENCODED_SHORT_CHANNEL_IDS],
        )


class ReplyShortChannelIDsMessage(Message):
    id = 262
    name = "reply_short_channel_ids"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.FULL_INFORMATION, SingleByteElement),
        ]


class QueryChannelRangeMessage(Message):
    id = 263
    name = "query_channel_range"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.FIRST_BLOCK_NUM, U32Element),
            (MessageProperty.NUMBER_OF_BLOCKS, U32Element),
        ]

    @classmethod
    def create(cls, chain_hash: bytes, first_block_num: int, number_of_blocks: int) -> Self:
        return cls(
            263,
            "query_channel_range",
            {
                MessageProperty.TYPE: MessageTypeElement(263, "query_channel_range"),
                MessageProperty.CHAIN_HASH: ChainHashElement(chain_hash),
                MessageProperty.FIRST_BLOCK_NUM: U32Element(first_block_num),
                MessageProperty.NUMBER_OF_BLOCKS: U32Element(number_of_blocks),
            },
        )

    @property
    def chain_hash(self):
        return cast(ChainHashElement, self.properties[MessageProperty.CHAIN_HASH])

    @property
    def first_block_num(self):
        return cast(U32Element, self.properties[MessageProperty.FIRST_BLOCK_NUM])

    @property
    def number_of_blocks(self):
        return cast(U32Element, self.properties[MessageProperty.NUMBER_OF_BLOCKS])


class ReplyChannelRangeMessage(Message):
    id = 264
    name = "reply_channel_range"

    @classmethod
    def features(cls) -> List[KeyedElement]:
        return super().features() + [
            (MessageProperty.CHAIN_HASH, ChainHashElement),
            (MessageProperty.FIRST_BLOCK_NUM, U32Element),
            (MessageProperty.NUMBER_OF_BLOCKS, U32Element),
            (MessageProperty.SYNC_COMPLETE, SingleByteElement),
            (MessageProperty.ENCODED_SHORT_CHANNEL_IDS, EncodedShortChannelIdsElement),
        ]

    @property
    def chain_hash(self):
        return cast(ChainHashElement, self.properties[MessageProperty.CHAIN_HASH])

    @property
    def first_block_num(self):
        return cast(U32Element, self.properties[MessageProperty.FIRST_BLOCK_NUM])

    @property
    def number_of_blocks(self):
        return cast(U32Element, self.properties[MessageProperty.NUMBER_OF_BLOCKS])

    @property
    def sync_complete(self):
        return cast(SingleByteElement, self.properties[MessageProperty.SYNC_COMPLETE])

    @property
    def encoded_short_channel_ids(self):
        return cast(
            EncodedShortChannelIdsElement,
            self.properties[MessageProperty.ENCODED_SHORT_CHANNEL_IDS],
        )

File: src/messages.rs
use crate::{
    node::Node,
    serialization::{
        ChainHashElement, FeaturesElement, GlobalFeaturesElement, IgnoredBytesElement,
        LocalFeaturesStruct, MessageTypeElement, NodeAddressesElement, NodeAliasElement,
        NumPongBytesElement, PointElement, SerializableToBytes, SerializationError,
        ShortChannelIDElement, SignatureElement, TLVStreamElement, TimestampElement,
        TimestampRangeElement, Wire1Byte, Wire3Bytes, WireU16Int, WireU16SizedBytes, WireU32Int,
        WireU64Int,
    },
};

use num_enum::TryFromPrimitive;
use strum::IntoEnumIterator;
use strum_macros::{EnumIter, IntoStaticStr};

#[derive(Debug, EnumIter, Copy, Clone, Eq, PartialEq, Hash, IntoStaticStr, TryFromPrimitive)]
#[repr(u16)]
pub enum MessageType {
    Unknown = 0,
    Warning = 1,
    Stfu = 2,
    // connection and keep alive
    Init = 16,
    Error = 17,
    Ping = 18,
    Pong = 19,
    // channel
    OpenChannel = 32,
    AcceptChannel = 33,
    FundingCreated = 34,
    FundingSigned = 35,
    ChannelReady = 36,
    Shutdown = 38,
    ClosingSigned = 39,
    ClosingComplete = 40,
    ClosingSig = 41,
    OpenChannel2 = 64,
    AcceptChannel2 = 65,
    // tx updates
    TxAddInput = 66,
    TxAddOutput = 67,
    TxRemoveInput = 68,
    TxRemoveOutput = 69,
    TxComplete = 70,
    TxSignatures = 71,
    TxInitRbf = 72,
    TxAckRbf = 73,
    TxAbort = 74,
    // channel updates and htlc management
    UpdateAddHTLC = 128,
    UpdateFulfillHTLC = 130,
    UpdateFailHTLC = 131,
    CommitmentSigned = 132,
    RevokeAndAck = 133,
    UpdateFee = 134,
    UpdateFailMalformedHTLC = 135,
    ChannelReestablish = 136,
    // network announcements
    ChannelAnnouncement = 256,
    NodeAnnouncement = 257,
    ChannelUpdate = 258,
    AnnouncementSignatures = 259,
    // gossip
    QueryShortChannelIds = 261,
    ReplyShortChannelIdsEnd = 262,
    QueryChannelRange = 263,
    ReplyChannelRange = 264,
    GossipTimestampFilter = 265,
}

impl MessageType {
    pub fn as_u16(&self) -> u16 {
        *self as u16
    }

    pub fn from_int(n: u16) -> Option<Self> {
        MessageType::iter().find(|&variant| variant as u16 == n)
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InitMessage {
    global_features: Vec<u8>,
    local_features: Vec<u8>,
    tlv: Vec<u8>,
}

impl SerializableToBytes for InitMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_message, data) = MessageTypeElement::from_bytes(data)?;
        let (global_features, data) = GlobalFeaturesElement::from_bytes(data)?;
        let (local_features, data) = LocalFeaturesStruct::from_bytes(data)?;
        let (tlv, data) = TLVStreamElement::from_bytes(data)?;
        Ok((
            InitMessage {
                global_features: global_features.value,
                local_features: local_features.value,
                tlv: tlv.value,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::Init).to_bytes());
        bytes.extend(GlobalFeaturesElement::new(self.global_features.clone()).to_bytes());
        bytes.extend(LocalFeaturesStruct::new(self.local_features.clone()).to_bytes());
        bytes.extend(TLVStreamElement::new(self.tlv.clone()).to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct PingMessage {
    pub num_pong_bytes: u16,
    pub ignored: IgnoredBytesElement,
}

impl SerializableToBytes for PingMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_message, data) = MessageTypeElement::from_bytes(data)?;
        let (num_pong_bytes, data) = NumPongBytesElement::from_bytes(data)?;
        let (ignored, data) = IgnoredBytesElement::from_bytes(data)?;
        Ok((
            PingMessage {
                num_pong_bytes: num_pong_bytes.value,
                ignored,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::Ping).to_bytes());
        bytes.extend(NumPongBytesElement::new(self.num_pong_bytes).to_bytes());
        bytes.extend(self.ignored.to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct PongMessage {
    ignored: IgnoredBytesElement,
}

impl PongMessage {
    pub fn from_ping(ping: PingMessage) -> Self {
        PongMessage {
            ignored: IgnoredBytesElement::new(vec![0; ping.num_pong_bytes as usize]),
        }
    }
}

impl SerializableToBytes for PongMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_message, data) = MessageTypeElement::from_bytes(data)?;
        let (ignored, data) = IgnoredBytesElement::from_bytes(data)?;
        Ok((PongMessage { ignored }, data))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::Pong).to_bytes());
        bytes.extend(self.ignored.to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct ChannelAnnouncementMessage {
    node_signature_1: SignatureElement,
    node_signature_2: SignatureElement,
    bitcoin_signature_1: SignatureElement,
    bitcoin_signature_2: SignatureElement,
    features: Vec<u8>,
    chain_hash: ChainHashElement,
    short_channel_id: ShortChannelIDElement,
    node_id_1: PointElement,
    node_id_2: PointElement,
    bitcoin_node_id_1: PointElement,
    bitcoin_node_id_2: PointElement,
}

impl SerializableToBytes for ChannelAnnouncementMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_message, data) = MessageTypeElement::from_bytes(data)?;
        let (node_signature_1, data) = SignatureElement::from_bytes(data)?;
        let (node_signature_2, data) = SignatureElement::from_bytes(data)?;
        let (bitcoin_signature_1, data) = SignatureElement::from_bytes(data)?;
        let (bitcoin_signature_2, data) = SignatureElement::from_bytes(data)?;
        let (features, data) = FeaturesElement::from_bytes(data)?;
        let (chain_hash, data) = ChainHashElement::from_bytes(data)?;
        let (short_channel_id, data) = ShortChannelIDElement::from_bytes(data)?;
        let (node_id_1, data) = PointElement::from_bytes(data)?;
        let (node_id_2, data) = PointElement::from_bytes(data)?;
        let (bitcoin_node_id_1, data) = PointElement::from_bytes(data)?;
        let (bitcoin_node_id_2, data) = PointElement::from_bytes(data)?;

        Ok((
            ChannelAnnouncementMessage {
                node_signature_1,
                node_signature_2,
                bitcoin_signature_1,
                bitcoin_signature_2,
                features: features.value,
                chain_hash,
                short_channel_id,
                node_id_1,
                node_id_2,
                bitcoin_node_id_1,
                bitcoin_node_id_2,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::ChannelAnnouncement).to_bytes());
        bytes.extend(self.node_signature_1.to_bytes());
        bytes.extend(self.node_signature_2.to_bytes());
        bytes.extend(self.bitcoin_signature_1.to_bytes());
        bytes.extend(self.bitcoin_signature_2.to_bytes());
        bytes.extend(FeaturesElement::new(self.features.clone()).to_bytes());
        bytes.extend(self.chain_hash.to_bytes());
        bytes.extend(self.short_channel_id.to_bytes());
        bytes.extend(self.node_id_1.to_bytes());
        bytes.extend(self.node_id_2.to_bytes());
        bytes.extend(self.bitcoin_node_id_1.to_bytes());
        bytes.extend(self.bitcoin_node_id_2.to_bytes());
        bytes
    }
}

#[derive(Debug, Clone)]
pub struct GossipTimestampFilterMessage {
    pub chain_hash: ChainHashElement,
    pub first_timestamp: u32,
    pub timestamp_range: u32,
}

impl SerializableToBytes for GossipTimestampFilterMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_, data) = MessageTypeElement::from_bytes(data)?;
        let (chain_hash, data) = ChainHashElement::from_bytes(data)?;
        let (first_timestamp, data) = TimestampElement::from_bytes(data)?;
        let (timestamp_range, data) = TimestampRangeElement::from_bytes(data)?;

        Ok((
            GossipTimestampFilterMessage {
                chain_hash,
                first_timestamp: first_timestamp.value,
                timestamp_range: timestamp_range.value,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::GossipTimestampFilter).to_bytes());
        bytes.extend(self.chain_hash.to_bytes());
        bytes.extend(TimestampElement::new(self.first_timestamp).to_bytes());
        bytes.extend(TimestampRangeElement::new(self.timestamp_range).to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct QueryChannelRangeMessage {
    chain_hash: ChainHashElement,
    first_blocknum: u32,
    number_of_blocks: u32,
    query_range_tlvs: Vec<u8>,
}

impl SerializableToBytes for QueryChannelRangeMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_, data) = MessageTypeElement::from_bytes(data)?;
        let (chain_hash, data) = ChainHashElement::from_bytes(data)?;
        let (first_blocknum, data) = WireU32Int::from_bytes(data)?;
        let (number_of_blocks, data) = WireU32Int::from_bytes(data)?;
        let (query_range_tlvs, data) = TLVStreamElement::from_bytes(data)?;

        Ok((
            QueryChannelRangeMessage {
                chain_hash,
                first_blocknum: first_blocknum.value,
                number_of_blocks: number_of_blocks.value,
                query_range_tlvs: query_range_tlvs.value,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::QueryChannelRange).to_bytes());
        bytes.extend(self.chain_hash.to_bytes());
        bytes.extend(WireU32Int::new(self.first_blocknum).to_bytes());
        bytes.extend(WireU32Int::new(self.number_of_blocks).to_bytes());
        bytes.extend(TLVStreamElement::new(self.query_range_tlvs.clone()).to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct ReplyChannelRangeMessage {
    chain_hash: ChainHashElement,
    first_blocknum: u32,
    number_of_blocks: u32,
    sync_complete: u8,
    encoded_short_ids: Vec<u8>,
    reply_channel_range_tlvs: Vec<u8>,
}

impl SerializableToBytes for ReplyChannelRangeMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_, data) = MessageTypeElement::from_bytes(data)?;
        let (chain_hash, data) = ChainHashElement::from_bytes(data)?;
        let (first_blocknum, data) = WireU32Int::from_bytes(data)?;
        let (number_of_blocks, data) = WireU32Int::from_bytes(data)?;
        let (sync_complete, data) = Wire1Byte::from_bytes(data)?;
        let (encoded_short_ids, data) = WireU16SizedBytes::from_bytes(data)?;
        let (reply_channel_range_tlvs, data) = TLVStreamElement::from_bytes(data)?;

        Ok((
            ReplyChannelRangeMessage {
                chain_hash,
                first_blocknum: first_blocknum.value,
                number_of_blocks: number_of_blocks.value,
                sync_complete: sync_complete.value,
                encoded_short_ids: encoded_short_ids.value,
                reply_channel_range_tlvs: reply_channel_range_tlvs.value,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::ReplyChannelRange).to_bytes());
        bytes.extend(self.chain_hash.to_bytes());
        bytes.extend(WireU32Int::new(self.first_blocknum).to_bytes());
        bytes.extend(WireU32Int::new(self.number_of_blocks).to_bytes());
        bytes.extend(Wire1Byte::new(self.sync_complete).to_bytes());
        bytes.extend(WireU16SizedBytes::new(self.encoded_short_ids.clone()).to_bytes());
        bytes.extend(TLVStreamElement::new(self.reply_channel_range_tlvs.clone()).to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct NodeAnnouncementMessage {
    signature: SignatureElement,
    features: Vec<u8>,
    timestamp: u32,
    pub node_id: PointElement,
    rgb_color: [u8; 3],
    alias: NodeAliasElement,
    addresses: NodeAddressesElement,
}

impl NodeAnnouncementMessage {
    pub fn as_node(&self) -> Option<Node> {
        let ipv4addr = match self.addresses.ipv4_addresses.first() {
            Some(ipv4addr) => ipv4addr,
            None => return None,
        };
        let ip_address = format!(
            "{}.{}.{}.{}",
            ipv4addr[0], ipv4addr[1], ipv4addr[2], ipv4addr[3]
        );
        let port = u16::from_be_bytes([ipv4addr[4], ipv4addr[5]]);
        Some(Node {
            public_key: self.node_id.value,
            ip_address,
            port,
        })
    }
}

impl SerializableToBytes for NodeAnnouncementMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_, data) = MessageTypeElement::from_bytes(data)?;
        let (signature, data) = SignatureElement::from_bytes(data)?;
        let (features, data) = WireU16SizedBytes::from_bytes(data)?;
        let (timestamp, data) = WireU32Int::from_bytes(data)?;
        let (node_id, data) = PointElement::from_bytes(data)?;
        let (rgb_color, data) = Wire3Bytes::from_bytes(data)?;
        let (alias, data) = NodeAliasElement::from_bytes(data)?;
        let (addresses, data) = NodeAddressesElement::from_bytes(data)?;

        Ok((
            NodeAnnouncementMessage {
                signature,
                features: features.value,
                timestamp: timestamp.value,
                node_id,
                rgb_color: rgb_color.value,
                alias,
                addresses,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::NodeAnnouncement).to_bytes());
        bytes.extend(self.signature.to_bytes());
        bytes.extend(WireU16SizedBytes::new(self.features.clone()).to_bytes());
        bytes.extend(WireU32Int::new(self.timestamp).to_bytes());
        bytes.extend(self.node_id.to_bytes());
        bytes.extend(Wire3Bytes::new(self.rgb_color).to_bytes());
        bytes.extend(self.alias.to_bytes());
        bytes.extend(self.addresses.to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct ChannelUpdateMessage {
    signature: SignatureElement,
    chain_hash: ChainHashElement,
    short_channel_id: ShortChannelIDElement,
    timestamp: u32,
    message_flags: u8,
    channel_flags: u8,
    cltv_expiry_delta: u16,
    htlc_minimum_msat: u64,
    fee_base_msat: u32,
    fee_proportional_millionths: u32,
    htlc_maximum_msat: u64,
}

impl SerializableToBytes for ChannelUpdateMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (_, data) = MessageTypeElement::from_bytes(data)?;
        let (signature, data) = SignatureElement::from_bytes(data).unwrap();
        let (chain_hash, data) = ChainHashElement::from_bytes(data).unwrap();
        let (short_channel_id, data) = ShortChannelIDElement::from_bytes(data).unwrap();
        let (timestamp, data) = TimestampElement::from_bytes(data).unwrap();
        let (message_flags, data) = Wire1Byte::from_bytes(data).unwrap();
        let (channel_flags, data) = Wire1Byte::from_bytes(data).unwrap();
        let (cltv_expiry_delta, data) = WireU16Int::from_bytes(data).unwrap();
        let (htlc_minimum_msat, data) = WireU64Int::from_bytes(data).unwrap();
        let (fee_base_msat, data) = WireU32Int::from_bytes(data).unwrap();
        let (fee_proportional_millionths, data) = WireU32Int::from_bytes(data).unwrap();
        let (htlc_maximum_msat, data) = WireU64Int::from_bytes(data).unwrap();

        Ok((
            ChannelUpdateMessage {
                signature,
                chain_hash,
                short_channel_id,
                timestamp: timestamp.value,
                message_flags: message_flags.value,
                channel_flags: channel_flags.value,
                cltv_expiry_delta: cltv_expiry_delta.value,
                htlc_minimum_msat: htlc_minimum_msat.value,
                fee_base_msat: fee_base_msat.value,
                fee_proportional_millionths: fee_proportional_millionths.value,
                htlc_maximum_msat: htlc_maximum_msat.value,
            },
            data,
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement::new(MessageType::ChannelUpdate).to_bytes());
        bytes.extend(self.signature.to_bytes());
        bytes.extend(self.chain_hash.to_bytes());
        bytes.extend(self.short_channel_id.to_bytes());
        bytes.extend(TimestampElement::new(self.timestamp).to_bytes());
        bytes.extend(Wire1Byte::new(self.message_flags).to_bytes());
        bytes.extend(Wire1Byte::new(self.channel_flags).to_bytes());
        bytes.extend(WireU16Int::new(self.cltv_expiry_delta).to_bytes());
        bytes.extend(WireU64Int::new(self.htlc_minimum_msat).to_bytes());
        bytes.extend(WireU32Int::new(self.fee_base_msat).to_bytes());
        bytes.extend(WireU32Int::new(self.fee_proportional_millionths).to_bytes());
        bytes.extend(WireU64Int::new(self.htlc_maximum_msat).to_bytes());
        bytes
    }
}

#[derive(Debug)]
pub struct UnknownMessage {
    type_id: u16,
    data: Vec<u8>,
}

impl SerializableToBytes for UnknownMessage {
    fn from_bytes(data: &[u8]) -> Result<(Self, &[u8]), SerializationError> {
        let (message, data) = MessageTypeElement::from_bytes(data)?;

        Ok((
            UnknownMessage {
                type_id: message.id,
                data: data.to_vec(),
            },
            &[],
        ))
    }

    fn to_bytes(&self) -> Vec<u8> {
        let mut bytes = Vec::new();
        bytes.extend(MessageTypeElement { id: self.type_id }.to_bytes());
        bytes.extend(self.data.clone());
        bytes
    }
}

#[test]
fn test_decode_init_message() {
    let initial_bytes = hex::decode("001000021100000708a0880a8a59a1012006226e46111a0b59caaf126043eb5bbf28c34f3a5e332a1fc7b2b73cf188910f2d7ef99482067a1b72fe9e411d37be8c").unwrap();
    let (msg, remainder) = InitMessage::from_bytes(&initial_bytes).unwrap();
    assert!(!msg.global_features.is_empty());
    assert!(!msg.local_features.is_empty());
    // check serialization
    assert_eq!([msg.to_bytes(), remainder.to_vec()].concat(), initial_bytes);
}
